droid-api/src/main/java/uk/gov/nationalarchives/droid/internal/api/DroidAPI.java

```diff
@@ -243,18 +274,68 @@ private List<ApiResult> submitHttpIdentification(final URI uri, String extension
 
         final RequestIdentifier id = getRequestIdentifier(uri);
 
+        Map<HashAlgorithm, String> hashResults = generateHashResults(uri, this::getHttpHash);
 
         try (final HttpIdentificationRequest request = new HttpIdentificationRequest(metaData, id, httpClient)) {
             request.setExtension(extension);
             request.open(uri);
-            return getApiResults(request);
+            return List.of(new APIResult(getIdentificationResults(request), hashResults));
         }
     }
 
-    private List<ApiResult> submitS3Identification(final URI uri, String extension) throws IOException {
+    private <T> Map<HashAlgorithm, String> generateHashResults(T identifier, BiFunction<HashAlgorithm, T, String> hashFunction) {
+        return hashAlgorithms.stream().collect(Collectors.toMap(
+                algorithm -> algorithm,
+                algorithm -> hashFunction.apply(algorithm, identifier)
+        ));
+    }
+
+    private String getFileHash(HashAlgorithm hashAlgorithm, Path path) {
+        try {
+            return getHash(hashAlgorithm, new FileInputStream(path.toFile()));
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String getS3Hash(HashAlgorithm algorithm, S3Uri s3Uri) {
+        String key = s3Uri.key().orElseThrow(() -> new RuntimeException("Key not found in uri " + s3Uri.uri()));
+        String bucket = s3Uri.bucket().orElseThrow(() -> new RuntimeException("Bucket not found in uri " + s3Uri.uri()));
+        ResponseInputStream<GetObjectResponse> responseInputStream = s3Client.getObject(GetObjectRequest.builder().bucket(bucket).key(key).build());
```

Comment:
similar to above, try-with-resource to close the `ResponseInputStream`

Reply:
I've changed this one as well.